service/migration_listener.hh

```diff
@@ -74,25 +80,43 @@ public:
     virtual void on_drop_aggregate(const sstring& ks_name, const sstring& aggregate_name) = 0;
     virtual void on_drop_view(const sstring& ks_name, const sstring& view_name) = 0;
 
+    // The callback runs inside seastar thread
+    // called before adding/updating/dropping column family. 
+    // listener can add additional type altering mutations if he knows what he is doing. 
+    virtual void on_before_create_column_family(const schema&, std::vector<mutation>&, api::timestamp_type) {}
+    virtual void on_before_update_column_family(const schema& new_schema, const schema& old_schema, std::vector<mutation>&, api::timestamp_type) {}
+    virtual void on_before_drop_column_family(const schema&, std::vector<mutation>&, api::timestamp_type) {}
+
     class only_view_notifications;
+    class empty_listener;
 };
 
 class migration_listener::only_view_notifications : public migration_listener {
-    virtual void on_create_keyspace(const sstring& ks_name) { }
-    virtual void on_create_column_family(const sstring& ks_name, const sstring& cf_name) { }
-    virtual void on_create_user_type(const sstring& ks_name, const sstring& type_name) { }
-    virtual void on_create_function(const sstring& ks_name, const sstring& function_name) { }
-    virtual void on_create_aggregate(const sstring& ks_name, const sstring& aggregate_name) { }
-    virtual void on_update_keyspace(const sstring& ks_name) { }
-    virtual void on_update_column_family(const sstring& ks_name, const sstring& cf_name, bool columns_changed) { }
-    virtual void on_update_user_type(const sstring& ks_name, const sstring& type_name) { }
-    virtual void on_update_function(const sstring& ks_name, const sstring& function_name) { }
-    virtual void on_update_aggregate(const sstring& ks_name, const sstring& aggregate_name) { }
-    virtual void on_drop_keyspace(const sstring& ks_name) { }
-    virtual void on_drop_column_family(const sstring& ks_name, const sstring& cf_name) { }
-    virtual void on_drop_user_type(const sstring& ks_name, const sstring& type_name) { }
-    virtual void on_drop_function(const sstring& ks_name, const sstring& function_name) { }
-    virtual void on_drop_aggregate(const sstring& ks_name, const sstring& aggregate_name) { }
+public:
+    void on_create_keyspace(const sstring& ks_name) override {}
```

Comment:
In my interpretation of our code style we use virtual for declarations (i.e. first occurrence in hierarchy), and override for overrides/impls. So yes, only cleanup. Because I really hate double decoration.